packages/app-store/_utils/oauth/OAuthManager.ts

```diff
@@ -0,0 +1,451 @@
+ *
+ * For a recommended usage example, see Zoom VideoApiAdapter.ts
+ */
+import type { z } from "zod";
+
+import { CREDENTIAL_SYNC_ENDPOINT } from "@calcom/lib/constants";
+import logger from "@calcom/lib/logger";
+import { safeStringify } from "@calcom/lib/safeStringify";
+
+import type { AxiosLikeResponseToFetchResponse } from "./AxiosLikeResponseToFetchResponse";
+import type { OAuth2TokenResponseInDbWhenExistsSchema, OAuth2UniversalSchema } from "./universalSchema";
+import { OAuth2UniversalSchemaWithCalcomBackwardCompatibility } from "./universalSchema";
+
+const log = logger.getSubLogger({ prefix: ["app-store/_utils/oauth/OAuthManager"] });
+export const enum TokenStatus {
+  UNUSABLE_TOKEN_OBJECT,
+  UNUSABLE_ACCESS_TOKEN,
+  VALID,
+}
+
+type ResourceOwner =
+  | {
+      id: number | null;
+      type: "team";
+    }
+  | {
+      id: number | null;
+      type: "user";
+    };
+
+type FetchNewTokenObject = ({ refreshToken }: { refreshToken: string | null }) => Promise<Response | null>;
+type UpdateTokenObject = (
+  token: z.infer<typeof OAuth2UniversalSchemaWithCalcomBackwardCompatibility>
+) => Promise<void>;
+type isTokenObjectUnusable = (response: Response) => Promise<{ reason: string } | null>;
+type isAccessTokenUnusable = (response: Response) => Promise<{ reason: string } | null>;
+type IsTokenExpired = (token: z.infer<typeof OAuth2UniversalSchema>) => Promise<boolean> | boolean;
+type InvalidateTokenObject = () => Promise<void>;
+type ExpireAccessToken = () => Promise<void>;
+type CredentialSyncVariables = {
+  /**
+   * The secret required to access the credential sync endpoint
+   */
+  CREDENTIAL_SYNC_SECRET: string | undefined;
+  /**
+   * The header name that the secret should be passed in
+   */
+  CREDENTIAL_SYNC_SECRET_HEADER_NAME: string;
+  /**
+   * The endpoint where the credential sync should happen
+   */
+  CREDENTIAL_SYNC_ENDPOINT: string | undefined;
+
+  APP_CREDENTIAL_SHARING_ENABLED: boolean;
+};
+/**
+ * Manages OAuth2.0 tokens for an app and resourceOwner
+ */
+export class OAuthManager {
+  private currentTokenObject: z.infer<typeof OAuth2UniversalSchema>;
+  private resourceOwner: ResourceOwner;
+  private appSlug: string;
+  private fetchNewTokenObject: FetchNewTokenObject;
+  private updateTokenObject: UpdateTokenObject;
+  private isTokenObjectUnusable: isTokenObjectUnusable;
+  private isAccessTokenUnusable: isAccessTokenUnusable;
+  private isTokenExpired: IsTokenExpired;
+  private invalidateTokenObject: InvalidateTokenObject;
+  private expireAccessToken: ExpireAccessToken;
+  private credentialSyncVariables: CredentialSyncVariables;
+  private useCredentialSync: boolean;
+  private autoCheckTokenExpiryOnRequest: boolean;
+
+  constructor({
+    resourceOwner,
+    appSlug,
+    currentTokenObject,
+    fetchNewTokenObject,
+    updateTokenObject,
+    isTokenObjectUnusable,
+    isAccessTokenUnusable,
+    invalidateTokenObject,
+    expireAccessToken,
+    credentialSyncVariables,
```

Comment:
Since the credential syncing variables are taken from the `.env` file do we need to pass these in the constructor every time?

Reply:
I made this change mainly to be able to toggle credential sync status in unit tests. 
 
But I think it is fine also as I wanted OAuthManager to be completely unaware of Application logic. So, it doesn't know what env variables or where exactly these constants should be imported from.